File: src/main/java/org/mitallast/queue/rest/netty/WebSocketFrameHandler.java
package org.mitallast.queue.rest.netty;


import com.google.inject.Inject;
import io.netty.buffer.ByteBuf;
import io.netty.buffer.Unpooled;
import io.netty.channel.ChannelHandler;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.SimpleChannelInboundHandler;
import io.netty.handler.codec.http.websocketx.BinaryWebSocketFrame;
import io.netty.handler.codec.http.websocketx.TextWebSocketFrame;
import io.netty.handler.codec.http.websocketx.WebSocketFrame;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.mitallast.queue.common.json.JsonService;

@ChannelHandler.Sharable
public class WebSocketFrameHandler extends SimpleChannelInboundHandler<WebSocketFrame> {
    private final Logger logger = LogManager.getLogger();
    private final JsonService jsonService;

    @Inject
    public WebSocketFrameHandler(JsonService jsonService) {
        this.jsonService = jsonService;
    }

    @Override
    public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
        logger.info("channel registered: {}", ctx.channel());
        super.channelActive(ctx);
    }

    @Override
    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
        logger.info("channel inactive: {}", ctx.channel());
        super.channelInactive(ctx);
    }

    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
        logger.error("unexpected exception", cause);
        super.exceptionCaught(ctx, cause);
    }

    @Override
    protected void channelRead0(ChannelHandlerContext ctx, WebSocketFrame frame) throws Exception {
        if (frame instanceof TextWebSocketFrame) {
            String json = ((TextWebSocketFrame) frame).text();
            logger.info("received {}", ctx.channel(), json);
        } else if (frame instanceof BinaryWebSocketFrame) {
            BinaryWebSocketFrame binaryFrame = (BinaryWebSocketFrame) frame;
            ByteBuf content = binaryFrame.content();
            int len = content.readableBytes();
            byte[] bytes = new byte[len];
            content.readBytes(bytes);
            ctx.writeAndFlush(new BinaryWebSocketFrame(Unpooled.wrappedBuffer(bytes)), ctx.voidPromise());
        } else {
            throw new UnsupportedOperationException("unsupported frame type: " + frame.getClass().getSimpleName());
        }
    }
}

File: src/main/java/org/mitallast/queue/rest/netty/WebSocketFrameHandler.kt
package org.mitallast.queue.rest.netty

import io.netty.buffer.Unpooled
import io.netty.channel.*
import io.netty.handler.codec.http.websocketx.BinaryWebSocketFrame
import io.netty.handler.codec.http.websocketx.TextWebSocketFrame
import io.netty.handler.codec.http.websocketx.WebSocketFrame
import io.vavr.collection.HashMap
import io.vavr.collection.Map
import org.mitallast.queue.common.logging.LoggingService

@ChannelHandler.Sharable
class WebSocketFrameHandler(logging: LoggingService) : SimpleChannelInboundHandler<WebSocketFrame>() {
    private val logger = logging.logger()

    @Volatile
    private var channels : Map<ChannelId, Channel> = HashMap.empty()

    @Throws(Exception::class)
    override fun channelRegistered(ctx: ChannelHandlerContext) {
        logger.info("channel registered: {}", ctx.channel())
        synchronized(this) {
            channels = channels.put(ctx.channel().id(), ctx.channel())
        }
        super.channelActive(ctx)
    }

    @Throws(Exception::class)
    override fun channelInactive(ctx: ChannelHandlerContext) {
        logger.info("channel inactive: {}", ctx.channel())
        synchronized(this) {
            channels = channels.remove(ctx.channel().id())
        }
        super.channelInactive(ctx)
    }

    @Throws(Exception::class)
    override fun channelRead0(ctx: ChannelHandlerContext, frame: WebSocketFrame) {
        when (frame) {
            is TextWebSocketFrame -> {
                val json = frame.text()
                logger.info("received {}", ctx.channel(), json)
            }
            is BinaryWebSocketFrame -> {
                val currentId = ctx.channel().id()
                val content = frame.content()
                val len = content.readableBytes()
                val bytes = ByteArray(len)
                content.readBytes(bytes)

                channels.values()
                    .filter { it.id() != currentId }
                    .forEach { it.writeAndFlush(BinaryWebSocketFrame(Unpooled.wrappedBuffer(bytes)), it.voidPromise()) }
            }
            else -> throw UnsupportedOperationException("unsupported frame type: " + frame.javaClass.simpleName)
        }
    }
}